src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpDownloadHandler.java

```diff
@@ -0,0 +1,135 @@
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.internal.StringUtil;
+import java.io.IOException;
+import java.io.OutputStream;
+
+class HttpDownloadHandler extends AbstractHttpHandler<HttpObject> {
+
+  private OutputStream out;
+  private boolean keepAlive = HttpVersion.HTTP_1_1.isKeepAliveDefault();
+
+  public HttpDownloadHandler(Credentials credentials) {
+    super(credentials);
+  }
+
+  @Override
+  protected void channelRead0(ChannelHandlerContext ctx, HttpObject msg) throws Exception {
+    checkState(userPromise != null, "response before request");
+    if (msg instanceof HttpResponse) {
+      HttpResponse response = (HttpResponse) msg;
+      keepAlive = HttpUtil.isKeepAlive((HttpResponse) msg);
+      if (!response.status().equals(HttpResponseStatus.OK)) {
+        failAndReset(
+            new HttpException(
+                response.status(), "Download failed with Status: " + response.status(), null),
+            ctx);
+      }
+    } else if (msg instanceof HttpContent) {
+      ByteBuf content = ((HttpContent) msg).content();
+      content.readBytes(out, content.readableBytes());
+      if (msg instanceof LastHttpContent) {
+        succeedAndReset(ctx);
+      }
+    } else {
+      failAndReset(
+          new IllegalArgumentException(
+              "Unsupported message type: " + StringUtil.simpleClassName(msg)),
+          ctx);
+    }
+  }
+
+  @Override
+  public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+      throws Exception {
+    checkState(userPromise == null, "handler can't be shared between pipelines.");
+    userPromise = promise;
+    if (!(msg instanceof DownloadCommand)) {
+      failAndResetUserPromise(
+          new IllegalArgumentException(
+              "Unsupported message type: " + StringUtil.simpleClassName(msg)));
+      return;
+    }
+    out = ((DownloadCommand) msg).out();
+    HttpRequest request = buildRequest((DownloadCommand) msg);
+    addCredentialHeaders(request, ((DownloadCommand) msg).uri());
+    ctx.writeAndFlush(request)
+        .addListener(
+            (f) -> {
+              if (!f.isSuccess()) {
+                failAndReset(f.cause(), ctx);
+              }
+            });
+  }
+
+  private HttpRequest buildRequest(DownloadCommand request) {
+    HttpRequest httpRequest =
+        new DefaultFullHttpRequest(
+            HttpVersion.HTTP_1_1,
+            HttpMethod.GET,
+            constructPath(request.uri(), request.hash(), request.casDownload()));
+    httpRequest.headers().set(HttpHeaderNames.HOST, constructHost(request.uri()));
```

Comment:
I wonder whether this code works with SSL SNI.

Reply:
Should be fine *guessing* :P. The TLS SNI is independent of the encapsulated protocol and the TLS handshake is done *before* the HTTP request is sent.